[Woo POS][Cash & Receipts] Complete order paid by cash

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14602

## Description
This PR adds the logic to mark orders as completed when paid by cash via the POS. Only tackles part of the happy path for the moment. We still have not implemented:
- Error handling/change due
- Redirect to success view upon completion

## Changes: 
- Updates `PointOfSaleOrderController` with a new `collectCashPayment()` method
- Dispatches `OrderAction.updateOrder` when the in-flight order is marked as completed.
- Updates existing `FormattableAmountTextField` with new `.pos` style, so we can reuse it.

## Testing
1. Enable `.acceptCashForPointOfSale`
2. Process a POS transaction via tapping the `Cash Payment" button
3. Observe the built-in keyboard can be used
4. Enter any amount in the text field. This reuses the `FormattableAmountTextField` component which already takes into account locale an currency.
5. Tap `Mark payment as completed`. Observe the loading indicator and the button being disabled while loading.
6. Once done, close the POS, go back to the app's Orders list, and observe the new order is there, marked complete and with `Cash (Point of Sale)` message (this is TBD right now).

https://github.com/user-attachments/assets/4528a047-8eac-45e9-b7ae-980195e81a6f

## Feedback:
I have not implemented these bits yet, so any advice is appreciated:
- Does it feel adequate to have the collect payment logic within the Order Controller? I think we can add here the error handling as well as the different methods needed for validating the cash amounts.
- Currently the presentation of the success view relies entirely on `CardPresentPaymentEventDetails.paymentSuccess`, but this doesn't make much sense if we accept cash. I'm thinking we should intercept `PointOfSaleAggregateModel.publishPaymentMessages` with a new publisher for cash, so we can not only present the success view but also cancel payment intents when cash is selected. Thoughts?

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
send action for cash payment order complete

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Merge branch 'task/14603-add-cash-option-to-totals-view' into task/14602-mark-payment-button

iamgabrielma committed:
Use existing FormattableAmountTextField

This component used for cash payments in order creation already handles locales and currency

iamgabrielma committed:
Add POS styling to formattableAmountTextField component

iamgabrielma committed:
Make button loading while callback

iamgabrielma committed:
add TODO for amount validation

iamgabrielma committed:
update error handling and constants

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleOrderController.swift:
```diff
@@ -86,6 +89,35 @@ final class PointOfSaleOrderController: PointOfSaleOrderControllerProtocol {
+                                                 onCompletion: { result in
+                continuation.resume(with: result)
+            })
+            ServiceLocator.stores.dispatch(action)
```
Do we write tests for the controller? If yes, this one will likely need to be injected.

iamgabrielma replied:
Good point, I've injected stores here: https://github.com/woocommerce/woocommerce-ios/pull/14739/commits/60e4748dac77509c4bc233d31ea6c498334e1c3f and added a test for the error case. I left out of the PR other tests for the success case for now, as I need to mock the order update. I'll handle this next when dealing with payment state changes.

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -38,18 +51,9 @@ struct PointOfSaleCollectCashView: View {
-                    Task { @MainActor in
-                        await markComplete()
-                    }
+            FormattableAmountTextField(viewModel: textFieldViewModel, style: .pos)
```
Nice that this can be reused with slight tweaks 👍

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleOrderController.swift:
```diff
@@ -157,3 +189,14 @@ extension PointOfSaleInternalOrderState: Equatable {
+
+extension PointOfSaleOrderController {
+    enum Constants {
+        static let cashPaymentMethodID: String = "cash"
```
Should we use `PaymentGateway.Constants.cashOnDeliveryGatewayID` "cod" or did we decide to use a different method ID and title for POS?

iamgabrielma replied:
I wasn't sure of this, I've asked on p1735049172231059?thread_ts=1735030361.006769&cid=C025A8VV728-slack-C025A8VV728 and we're using `cod` ID and "Pay in Person" title indeed. Updated on https://github.com/woocommerce/woocommerce-ios/pull/14739/commits/37afc8c986453b44e859a4fcbe964bad5acd5697

jaclync replied:
Updated link: p1735049172231059/1735030361.006769-slack-C025A8VV728

staskus reviewed (approved):
Thanks for the work. Left a couple of questions and notes.

### UI notes for the future

**The view doesn't adapt great for larger font sizes**

<img width="1180" alt="image" src="https://github.com/user-attachments/assets/0745ace0-5739-4220-87fc-5c6ace526533" />

<img width="1180" alt="image" src="https://github.com/user-attachments/assets/00495d55-11ad-497a-b992-b50508bb9d83" />


**Background color is different between cash payment and other views**

|                         |                          |
|-------------------------|--------------------------|
| ![IMG_0140](https://github.com/user-attachments/assets/69eb66e4-b2e7-4029-8b4b-f2d9638accce) | ![IMG_0139](https://github.com/user-attachments/assets/90ab06bc-c8bb-4edc-8f02-fd10c6ec6c46) |

iamgabrielma committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-ios into task/14602-mark-payment-button

iamgabrielma committed:
Update cash payment method ID to COD

iamgabrielma committed:
Inject stores into order controller

iamgabrielma committed:
use secondary for cash modal bg color

iamgabrielma commented:
Thanks for the review! 

> The view doesn't adapt great for larger font sizes

True that, logged here: https://github.com/woocommerce/woocommerce-ios/issues/14792

> Background color is different between cash payment and other views

I've updated the full-screen modal background here: https://github.com/woocommerce/woocommerce-ios/pull/14739/commits/58801654ce4c4f2d87a403c1a88165a4258c106e . For the `TotalsView` and rest of subviews I think we have a design discrepancy, as the i1 iteration shows the same colour for all views but that's not true on the pre-existing design, even light mode has different tones. I'll check with design what we want to do with these both in light and dark modes:
<img width="492" alt="Screenshot 2025-01-06 at 12 21 29" src="https://github.com/user-attachments/assets/022973da-c24c-4875-abca-ad0186686751" />
